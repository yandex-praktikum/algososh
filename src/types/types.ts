import { ElementStates } from "./element-states";

export interface stringCharsProps {
  adding?: boolean;
  deleting?: boolean;
  noArrow?: boolean;
  tail?: string;
  head?: string;
  char?: string;
  state: ElementStates;
  extraCircle?: {
    char: string;
  }
}

export interface columnObject {
  num: number;
  state: ElementStates;
}

export type radioButtonState = "selection" | "bubble";
