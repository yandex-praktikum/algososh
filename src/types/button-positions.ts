export enum ButtonPositions {
    Top = 'Top',
    Bottom = 'Bottom',
  }
  