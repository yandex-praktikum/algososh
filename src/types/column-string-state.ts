export interface IStringColumnStateProps {
  index: number;
  steps: string[][];
  currentStepIndex: number;
}
