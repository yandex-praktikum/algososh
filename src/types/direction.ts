export enum Direction {
  Ascending = 'ascending',
  Descending = 'descending',
}
