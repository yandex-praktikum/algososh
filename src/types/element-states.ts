export enum ElementStates {
  Default = "default",
  Changing = "changing",
  Modified = "modified",
  Chosen = "chosen"
}
