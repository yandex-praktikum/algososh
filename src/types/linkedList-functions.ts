export enum Functions {
    Waiting = '',
    AddToHead = 'AddToHead',
    AddToTail = 'AddToTail',
    DeleteFromHead = 'DeleteFromHead',
    DeleteFromTail = 'DeleteFromTail',
    AddByIndex = 'AddByIndex',
    DeleteByIndex = 'DeleteByIndex',
  }
  