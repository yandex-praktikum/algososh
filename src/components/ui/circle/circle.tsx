import React from "react";
import styles from "./circle.module.css";
import { ElementStates } from "../../../types/element-states";

interface CircleProps {
  state?: ElementStates;
  letter?: string | null;
  head?: string | React.ReactElement | null;
  index?: number;
  tail?: string | React.ReactElement | null;
  tailType?: "string" | "element";
  extraClass?: string;
  isSmall?: boolean;
}

export const Circle: React.FC<CircleProps> = ({
  state = ElementStates.Default,
  letter,
  head,
  index,
  tail,
  extraClass = "",
  isSmall,
}) => {
  return (
    <div className={`${styles.content} ${extraClass}`}>
      <div
        className={`text text_type_input text_color_input mb-4 ${
          styles.absolute
        } ${styles.head} ${
          styles[typeof head === "string" ? "string" : "element"]
        }`}
      >
        {head}
      </div>
      <div
        className={`${styles.circle}  ${isSmall ? styles.small : ""} ${
          styles[state]
        }`}
      >
        <p
          className={`text text_type_circle text_color_input ${styles.letter}`}
        >
          {letter}
        </p>
      </div>
      <p
        className={`text text_type_input text_color_input mt-4 ${styles.absolute} ${styles.index}`}
      >
        {index?.toString()}
      </p>
      <div
        className={`text text_type_input text_color_input mt-4 ${
          styles.absolute
        } ${index?.toString() ? styles.tail60 : styles.tail30} ${
          styles[typeof tail === "string" ? "string" : "element"]
        }`}
      >
        {tail}
      </div>
    </div>
  );
};
