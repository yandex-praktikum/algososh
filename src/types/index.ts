import { Functions } from "./linkedList-functions";
import { Direction } from "./direction";
import { ElementStates } from "./element-states";
import { ButtonPositions } from "./button-positions";
import { SortingMethods } from "./sorting";
import type { TSortingResult, TReverseStringResult } from "./results";

export type { TSortingResult, TReverseStringResult };

export { Functions, Direction, ButtonPositions, ElementStates, SortingMethods };

export type TGetElementState = {
  itemIndex: number;
  startPosition: number;
  endPosition: number;
  isReversed: boolean;
  timerLaunched: boolean;
};

export type TSortingSteps<T> = {
  array: T[];
  current: number[];
  modified: number[];
}[];

export interface IRefStep {
  current?: number[];
  modified?: number;
}
