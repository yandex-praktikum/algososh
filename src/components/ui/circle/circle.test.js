import { Circle } from "./circle";
import React from "react";
import renderer from "react-test-renderer";
import { render, fireEvent, waitFor, screen } from "@testing-library/react";
import { ElementStates } from "../../../types/element-states";

describe("Проверка рендера кружка", () => {
  describe("Рендер с буквами и без букв", () => {
    it("Кружок без буквы рендерится без ошибок", () => {
      const tree = renderer.create(<Circle />).toJSON();
      expect(tree).toMatchSnapshot();
    });

    it("Кружок c буквами рендерится без ошибок", () => {
      const tree = renderer.create(<Circle letter="TEST" />).toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
  describe("Кружок рендериться с head", () => {
    it("Кружок cо строкой в head рендерится без ошибок", () => {
      const tree = renderer.create(<Circle head={"test"} />).toJSON();
      expect(tree).toMatchSnapshot();
    });
    it("Кружок c node в head рендерится без ошибок", () => {
      const tree = renderer.create(<Circle head={<Circle />} />).toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
  describe("Кружок рендериться с tail", () => {
    it("Кружок cо строкой в tail рендерится без ошибок", () => {
      const tree = renderer.create(<Circle tail={"test"} />).toJSON();
      expect(tree).toMatchSnapshot();
    });
    it("Кружок c node в tail рендерится без ошибок", () => {
      const tree = renderer.create(<Circle tail={<Circle />} />).toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
  describe("Кружок рендериться в разных состояниях", () => {
    it("Кружок в состоянии default рендерится без ошибок", () => {
      const tree = renderer
        .create(<Circle state={ElementStates.Default} />)
        .toJSON();
      expect(tree).toMatchSnapshot();
    });
    it("Кружок в состоянии changing рендерится без ошибок", () => {
      const tree = renderer
        .create(<Circle state={ElementStates.Changing} />)
        .toJSON();
      expect(tree).toMatchSnapshot();
    });
    it("Кружок в состоянии modified рендерится без ошибок", () => {
      const tree = renderer
        .create(<Circle state={ElementStates.Modified} />)
        .toJSON();
      expect(tree).toMatchSnapshot();
    });
    it("Кружок в состоянии chosen рендерится без ошибок", () => {
      const tree = renderer
        .create(<Circle state={ElementStates.Chosen} />)
        .toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
  it("Кружок с индексом рендерится без ошибок", () => {
    const tree = renderer.create(<Circle index={1} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
  it("Кружок с пропом isSmall рендерится без ошибок", () => {
    const tree = renderer.create(<Circle isSmall={true} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
