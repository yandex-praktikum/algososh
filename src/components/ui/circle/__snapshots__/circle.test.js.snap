// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Проверка рендера кружка Кружок рендериться в разных состояниях Кружок в состоянии changing рендерится без ошибок 1`] = `
<div
  className="content "
>
  <div
    className="text text_type_input text_color_input mb-4 absolute head element"
  />
  <div
    className="circle   changing"
  >
    <p
      className="text text_type_circle text_color_input letter"
    />
  </div>
  <p
    className="text text_type_input text_color_input mt-4 absolute index"
  />
  <div
    className="text text_type_input text_color_input mt-4 absolute tail30 element"
  />
</div>
`;

exports[`Проверка рендера кружка Кружок рендериться в разных состояниях Кружок в состоянии chosen рендерится без ошибок 1`] = `
<div
  className="content "
>
  <div
    className="text text_type_input text_color_input mb-4 absolute head element"
  />
  <div
    className="circle   chosen"
  >
    <p
      className="text text_type_circle text_color_input letter"
    />
  </div>
  <p
    className="text text_type_input text_color_input mt-4 absolute index"
  />
  <div
    className="text text_type_input text_color_input mt-4 absolute tail30 element"
  />
</div>
`;

exports[`Проверка рендера кружка Кружок рендериться в разных состояниях Кружок в состоянии default рендерится без ошибок 1`] = `
<div
  className="content "
>
  <div
    className="text text_type_input text_color_input mb-4 absolute head element"
  />
  <div
    className="circle   default"
  >
    <p
      className="text text_type_circle text_color_input letter"
    />
  </div>
  <p
    className="text text_type_input text_color_input mt-4 absolute index"
  />
  <div
    className="text text_type_input text_color_input mt-4 absolute tail30 element"
  />
</div>
`;

exports[`Проверка рендера кружка Кружок рендериться в разных состояниях Кружок в состоянии modified рендерится без ошибок 1`] = `
<div
  className="content "
>
  <div
    className="text text_type_input text_color_input mb-4 absolute head element"
  />
  <div
    className="circle   modified"
  >
    <p
      className="text text_type_circle text_color_input letter"
    />
  </div>
  <p
    className="text text_type_input text_color_input mt-4 absolute index"
  />
  <div
    className="text text_type_input text_color_input mt-4 absolute tail30 element"
  />
</div>
`;

exports[`Проверка рендера кружка Кружок рендериться с head Кружок c node в head рендерится без ошибок 1`] = `
<div
  className="content "
>
  <div
    className="text text_type_input text_color_input mb-4 absolute head element"
  >
    <div
      className="content "
    >
      <div
        className="text text_type_input text_color_input mb-4 absolute head element"
      />
      <div
        className="circle   default"
      >
        <p
          className="text text_type_circle text_color_input letter"
        />
      </div>
      <p
        className="text text_type_input text_color_input mt-4 absolute index"
      />
      <div
        className="text text_type_input text_color_input mt-4 absolute tail30 element"
      />
    </div>
  </div>
  <div
    className="circle   default"
  >
    <p
      className="text text_type_circle text_color_input letter"
    />
  </div>
  <p
    className="text text_type_input text_color_input mt-4 absolute index"
  />
  <div
    className="text text_type_input text_color_input mt-4 absolute tail30 element"
  />
</div>
`;

exports[`Проверка рендера кружка Кружок рендериться с head Кружок cо строкой в head рендерится без ошибок 1`] = `
<div
  className="content "
>
  <div
    className="text text_type_input text_color_input mb-4 absolute head string"
  >
    test
  </div>
  <div
    className="circle   default"
  >
    <p
      className="text text_type_circle text_color_input letter"
    />
  </div>
  <p
    className="text text_type_input text_color_input mt-4 absolute index"
  />
  <div
    className="text text_type_input text_color_input mt-4 absolute tail30 element"
  />
</div>
`;

exports[`Проверка рендера кружка Кружок рендериться с tail Кружок c node в tail рендерится без ошибок 1`] = `
<div
  className="content "
>
  <div
    className="text text_type_input text_color_input mb-4 absolute head element"
  />
  <div
    className="circle   default"
  >
    <p
      className="text text_type_circle text_color_input letter"
    />
  </div>
  <p
    className="text text_type_input text_color_input mt-4 absolute index"
  />
  <div
    className="text text_type_input text_color_input mt-4 absolute tail30 element"
  >
    <div
      className="content "
    >
      <div
        className="text text_type_input text_color_input mb-4 absolute head element"
      />
      <div
        className="circle   default"
      >
        <p
          className="text text_type_circle text_color_input letter"
        />
      </div>
      <p
        className="text text_type_input text_color_input mt-4 absolute index"
      />
      <div
        className="text text_type_input text_color_input mt-4 absolute tail30 element"
      />
    </div>
  </div>
</div>
`;

exports[`Проверка рендера кружка Кружок рендериться с tail Кружок cо строкой в tail рендерится без ошибок 1`] = `
<div
  className="content "
>
  <div
    className="text text_type_input text_color_input mb-4 absolute head element"
  />
  <div
    className="circle   default"
  >
    <p
      className="text text_type_circle text_color_input letter"
    />
  </div>
  <p
    className="text text_type_input text_color_input mt-4 absolute index"
  />
  <div
    className="text text_type_input text_color_input mt-4 absolute tail30 string"
  >
    test
  </div>
</div>
`;

exports[`Проверка рендера кружка Кружок с индексом рендерится без ошибок 1`] = `
<div
  className="content "
>
  <div
    className="text text_type_input text_color_input mb-4 absolute head element"
  />
  <div
    className="circle   default"
  >
    <p
      className="text text_type_circle text_color_input letter"
    />
  </div>
  <p
    className="text text_type_input text_color_input mt-4 absolute index"
  >
    1
  </p>
  <div
    className="text text_type_input text_color_input mt-4 absolute tail60 element"
  />
</div>
`;

exports[`Проверка рендера кружка Кружок с пропом isSmall рендерится без ошибок 1`] = `
<div
  className="content "
>
  <div
    className="text text_type_input text_color_input mb-4 absolute head element"
  />
  <div
    className="circle  small default"
  >
    <p
      className="text text_type_circle text_color_input letter"
    />
  </div>
  <p
    className="text text_type_input text_color_input mt-4 absolute index"
  />
  <div
    className="text text_type_input text_color_input mt-4 absolute tail30 element"
  />
</div>
`;

exports[`Проверка рендера кружка Рендер с буквами и без букв Кружок c буквами рендерится без ошибок 1`] = `
<div
  className="content "
>
  <div
    className="text text_type_input text_color_input mb-4 absolute head element"
  />
  <div
    className="circle   default"
  >
    <p
      className="text text_type_circle text_color_input letter"
    >
      TEST
    </p>
  </div>
  <p
    className="text text_type_input text_color_input mt-4 absolute index"
  />
  <div
    className="text text_type_input text_color_input mt-4 absolute tail30 element"
  />
</div>
`;

exports[`Проверка рендера кружка Рендер с буквами и без букв Кружок без буквы рендерится без ошибок 1`] = `
<div
  className="content "
>
  <div
    className="text text_type_input text_color_input mb-4 absolute head element"
  />
  <div
    className="circle   default"
  >
    <p
      className="text text_type_circle text_color_input letter"
    />
  </div>
  <p
    className="text text_type_input text_color_input mt-4 absolute index"
  />
  <div
    className="text text_type_input text_color_input mt-4 absolute tail30 element"
  />
</div>
`;
